### ISSUE TITLE
Beta: isAuthenticated true after access and refresh tokens are expired

### ISSUE THREAD

**Issue by spenweb** [auth0/auth0-react#464]
### Describe the problem

`isAuthenticated` continues to be `true` when access and refresh tokens are expired. The client then starts making HTTP calls to protected API endpoints because it is expecting a valid access token to be available. However, the client starts receiving HTTP 403 errors because the session is NOT authenticated. 

### What was the expected behavior?

I would expect that `isAuthenticated` would be set to `false` after both the refresh and access tokens have expired.

### Reproduction

Clone and follow the readme in this demo NextJS app I created to reproduce the issue: https://github.com/spenweb/auth-for-real

**Can the behavior be reproduced using the [React SDK Playground](https://github.com/auth0/auth0-react/blob/master/CONTRIBUTING.md#local-development)?**

### Environment

- **Version of `auth0-react` used:** ^2.0.0-beta.0
- **Which browsers have you tested in?** Chrome (108.0.5359.124)
- **Which framework are you using, if applicable (Angular, React, etc):** NextJS (next@13.1.1)
- **Other modules/plugins/libraries that might be involved:** Check demo repo package.json

**Comment by adamjmcgrath** [comment]
Hi @spenweb - thanks for raising this 

Have a look at https://github.com/auth0/auth0-spa-js/issues/1049 for details on the different behaviour between v1 and v2

**Comment by ewanharris** [comment]
Hey @spenweb, what you're seeing here is down to some changes within SPA-JS v2 around how we treat ID tokens and access tokens.

In v1 we determine the logged in state off the access token(s), so if you have multiple access tokens for multiple audiences you would need to specify an audience to determine if you’re logged in, this doesn’t work well with auth0-react since there’s only one isAuthenticated prop.

In v2 we base the logged in state off of the ID token, so even if your access token is expired, you’re still logged in until you logout.

This is more inline with how we view a SPA’s “session” and more flexible for users of multiple audiences, but means that you need to handle failed refresh token exchanges in the same way you would handle 401's from your API, by logging the user in again (or just logging them out and prompting them to login again).

So in your sample app, this could lead to the `getAccessTokenSilently` call looking as follows:

```js
try {
    const token = await getAccessTokenSilently();
    setToken(token);
} catch (e) {
    // Handle an expired token by redirecting the user to login
    if (e.error === 'invalid_grant') {
        loginWithRedirect({
            authorizationParams: { prompt: "login" },
        });
        return;
    }
    console.error(e);
    setToken("");
}
```

**Comment by spenweb** [comment]
Thanks for the feedback @ewanharris. Is this documented anywhere yet? 

Also, even when the ID Token was expired, `isAuthenticated` was set to true. Not sure how to handle that.

**Comment by ewanharris** [comment]
@spenweb, I've made a note for us to ensure this gets documented in the migration guide ahead of the GA release.

The expiry of the ID token is unrelated to a session as it is to be used when verifying the ID token on the client, this is why an ID token will generally also have a shorter lifetime. As it's only for that verification you shouldn't need to worry about that

**Comment by spenweb** [comment]
> In v2 we base the logged in state off of the ID token, so even if your access token is expired, you’re still logged in until you logout.

When would `isAuthenticated` be set to `false`? It seems counterintuitive to me that the logged in state is based on the existence of an ID token but it does not take into account when that ID token is expired.

**Comment by frederikprijck** [comment]
In a SPA, all we can do is assume authentication state, unless we contact Auth0. Authentication state should never be done based on access tokens (as we can have multiple), and an ID token's expiration should also not be used for this. ID tokens are used to exchange user information, not to drive the SPA "session" duration.

The only real indication is when Auth0 tells us the user is not authenticated anymore, this should happen under the hood when Auth0 is contacted. If Auth0, at some point, returns a `login_required` error, we will ensure `isAuthenticated` returns false, see: https://github.com/auth0/auth0-spa-js/blob/master/src/Auth0Client.ts#L921

In general, once the user has authenticated, there is nothing we can use to know if the user isn't authenticated anymore, apart from contacting Auth0. The expiration of any such tokens doesn't mean the user is not authenticated anymore with Auth0, and even though the current tokens could be expired, we should be able to get a new access and id token by using `getAccessTokenSilently`.

Note, this can get complicated when you use refresh tokens, especially in v2, where we have no fallback to use iframes by default. This means, if the refresh token expired, the user is left with a situation where `isAuthenticated` will be true, yet they have no ability to get a new set of valid access tokens. Can you verify whether this is or isn't your situation @spenweb ?

Regardless, @ewanharris  I think we might need to look into if there is anything we can do to improve the situation where:

- Refresh tokens are used
- Iframe fallback is disabled (by default in v2)
- Refresh token is expired

The only way to get a new pair of valid tokens is to call `loginWithRedirect` again, making me wonder that `isAuthenticated` should have been false in this case.

**Comment by spenweb** [comment]
@frederikprijck 
> Note, this can get complicated when you use refresh tokens, especially in v2, where we have no fallback to use iframes by default. This means, if the refresh token expired, the user is left with a situation where isAuthenticated will be true, yet they have no ability to get a new set of valid access tokens. Can you verify whether this is or isn't your situation @spenweb ?

Yes, I can confirm this is happening using my [demo codebase](https://github.com/spenweb/auth-for-real).

![image](https://user-images.githubusercontent.com/46385475/210865979-c7108457-52d6-4516-bdeb-c9ca7f82b65e.png)

Here is my Auth0Provider setup:
```tsx
<Auth0Provider
  domain={NEXT_PUBLIC_AUTH0_DOMAIN}
  clientId={NEXT_PUBLIC_AUTH0_CLIENT_ID}
  authorizationParams={{
    audience: NEXT_PUBLIC_AUTH0_AUDIENCE,
    scope: NEXT_PUBLIC_AUTH0_SCOPE,
    redirect_uri: "http://localhost:3000",
  }}
  useRefreshTokens={true}
  cacheLocation="localstorage"
  useRefreshTokensFallback={false}
>
  {children}
</Auth0Provider>
```


> Even more so, can you verify things do work as expected when you set useRefreshTokensFallback to true?

Yes, they work as expected when setting useRefreshTokensFallback to true.

**Comment by spenweb** [comment]
> In general, once the user has authenticated, there is nothing we can use to know if the user isn't authenticated anymore, apart from contacting Auth0.

So `isAuthenticated` really is more like `isMaybeAuthenticatedToAuth0`. I was originally reading it more as `isAuthenticatedToMyAPIServer`. I am currently only using one audience (one API server) with Auth0. 

It would be awesome if there was a convenience method that was part of this package that allowed for checking whether the current session is authenticated for a particular audience. Also, if there was an event or callback triggered when the user session is no longer authenticated due to refresh token expiration when `useRefreshTokensFallback={false}`.

**Comment by frederikprijck** [comment]
> So isAuthenticated really is more like isMaybeAuthenticatedToAuth0. I was originally reading it more as isAuthenticatedToMyAPIServer. I am currently only using one audience (one API server) with Auth0.

You are never _authenticated_ to an API. You are **authorized** to an API. So yes, isAuthenticated has nothing to do with API's, access tokens and refresh tokens.

> It would be awesome if there was a convenience method that was part of this package that allowed for checking whether the current session is authenticated for a particular audience. Also, if there was an event or callback triggered when the user session is no longer authenticated due to refresh token expiration when useRefreshTokensFallback={false}.

Once you start using refresh tokens, you basically tell the IDP you want to use offline_access, meaning you want to be able to keep issuing access tokens despite the session with the IDP being active or not. Once you use refresh tokens, it makes no sense to think about `checking whether or not a session is active`, as you explicitly said you want to bypass that session using refresh tokens.

The only thing you can do is catch any call to `getAccessTokenSilently` and react when it throws an `invalid_grant` error. This means Auth0 can not process the refresh tokens, which could mean it's expired.

**Comment by tim5046** [comment]
I don't know that we should close this issue as completed. A prop like `isAuthenticated` should return true if authenticated or false if unauthenticated. I believe a more clear name like `isMaybeAuthenticated` would be appropriate here. I too fell prey to trusting that `isAuthenticated === true` meant something that it apparently doesn't. This is a particularly common problem with users on Safari, because the refresh tokens generally don't come through and the error needs to be caught. But in those cases when the refresh tokens fail, `isAuthenticated` remains true. This is confusing and I'd hope for a clearer name on the property.

**Comment by frederikprijck** [comment]
Hey @tim5046 , you use a refresh token to get a new access token. Access tokens have little to do with authentication, but more with authorization towards an api. U can be authenticated without having both an access token and a refresh token. Authentication is what the id token is for, whose expiration does not impact authenticated state.

**Comment by tim5046** [comment]
> Can you elaborate on what you mean with the refresh token not coming through on safari? It might make sense to open a seperate issue for that if you believe there is an issue with our SDK.

With Safari, Brave, anything with ITP enabled really, the refresh token behavior largely does not work well. We're catching errors all the time with Safari refresh tokens not actually refreshing and manually throwing the user to log out in the background. It's really rough

Here's a really silly function we have all over the place that just catches Safari refresh tokens and manually logs them back in (or out, depending on the case):
``` 
getAccessTokenSilently()
      .then((token) => token)
      .catch(() =>
        loginWithRedirect({
          authorizationParams: {
            redirect_uri: `${window.location.origin}/handshake`,
          },
        })
      );
```

I take your point on the technical accuracy of `isAuthenticated` from auth0's standpoint. Our team however lost a ton of time on it. We naively did code checks like:

```
if (isAuthenticated) {
  // Do something that sends the token to our API
  // only for that token to sometimes error because it's expired
} 
```

Now my alternative, as I understand it, is something like this:
```
const ensureIsAuthenticated = () => {
  getAccessTokenSilently()
      .catch(() => {
          setIsActuallyAuthenticated(true); <- use this local state variable now instead of auth0's isAuthenticated

          // Maybe log the user out or do something else here
    })
}
```

This just feels like a really suboptimal flow.

**Comment by spenweb** [comment]
Might be worth creating a wrapper around Auth0's hooks that handles the expired tokens. Then you can have your own `isAuthenticated` variable have your team's desired meaning of "the current user should be able to successfully call the API endpoint". Could even use the same TypeScript interfaces so you can just change the import statements to switch from using Auth0's SDK directly to your own wrapped hooks.

**Comment by frederikprijck** [comment]
Hey @tim5046 , are you using refresh tokens or custom domains ? Both of those should work around third party cookies and work in browsers with ITP enabled.

If u need help, let's move to a seperate issue, but it should work, see https://github.com/auth0/auth0-react/blob/master/FAQ.md#1-user-is-not-logged-in-after-page-refresh.

**Comment by frederikprijck** [comment]
Having said all of that, let's improve the docs to callout the meaning of isAuthenticated more clearly.

**Comment by tim5046** [comment]
Thank you **SO MUCH** Frederik.  We are using custom domains, but we have not set the `useRefreshTokensFallback={true}` prop. Are you saying that this will trigger a call to auth0 when a refresh token fails to refresh and thus `isAuthenticated` would return false in that case? If so that's a godsend. We've been struggling with auth token issues for a few weeks now

**Comment by Bartheleway** [comment]
It's urgent that Auth0 clarify that because it cause a lot of confusion especially when your own documentation state that login should be offered only if `isAuthenticated` returns false :
- https://auth0.com/docs/quickstart/spa/vanillajs/01-login#restoring-login-state-with-social-providers
- https://auth0.com/docs/quickstart/spa/vanillajs/interactive
- https://github.com/auth0-samples/auth0-javascript-samples/blob/d45978c40db91b30ef81abcd9114f784aff5234b/01-Login/public/js/app.js#L62
- https://auth0.github.io/auth0-spa-js/classes/Auth0Client.html#isAuthenticated

**Comment by jimniels** [comment]
FWIW my use case and finding this thread.

I was using `isAuthenticated` to allow people to access protected routes in the app (e.g. "are they logged in"). If it was `true`, then the protected routes would render.

But I found that when you'd come back to the app after a while, the "logged in" view of the app would render, but only for a split second and then you'd get redirected to Auth0.

I found this was because React would render the "logged in" view of the app and then, somewhere down in the tree it would try to call the API (in order to show user-specific data for the "logged in" view) at which point it would try to get an access token and then fail because the access token was old, and then it would `catch` and redirect the user to the login page.

I couldn't understand why someone would be "authenticated" and "get in to" the app but then immediately be redirected back to login. "Why would `isAuthenticated` be true, but then getting an access token would fail?"

Then I found this thread, which helped me understand that from auth0's point of view authentication !== authorization

In our case, when we want to know if someone can view a protected route (e.g. "are they logged in?") that is conceptually the same as "are they authorized?" because as soon as they are logged in, they are going to need to retrieve data from our API so they'll need a token.

So it sounds like what we really need is to just ignore the `isAuthenticated` flag and implement our our boolean like `isAuthorized`, which would be something like:

```
async isAuthorized() {
  const client = await getClient();
  try {
    await client.getTokenSilently();
    return true;
  } catch (e) {
    return false;
  }
}
```

Where we can say "is this person logged in to the app, and therefore do they have the ability to make API calls?" If so, let them have access to protected routes (which call the API). Otherwise, they are "logged out" of the app and shouldn't have access to protected routes.

---

If what I have described above sounds right, I think it would be useful for there to be extra clarification in the docs about what `isAuthenticated` means, why you would use it, and how if you just want to know "is this person logged in _and therefore_ do they have an access token for making API calls?" what the convention is for that (as it seems like a common conceptual idea people would want?)

**Comment by frederikprijck** [comment]
Something like that could make sense in your own application if you need some concept of isAuthorized. However, it can become more complicated as the question might become: "Authorized for what"? You can have all kinds of combinations of audiences and scopes, that all result in being authorized for other things.

E.g. you can be authorized for `scope="write:messages"` on `APIX`, and `scope="read:logs"` on `APIY`, which is why getTokenSilently accepts an audience and scope param in the first place. 

Sure, you can call it isAuthorized, but in essence you are wrapping `getTokenSIlently` and could just forget about the concept of being authorized and just think in terms of "do I have a token to do what I want", and rely on our `getTokenSilently` method.

Regarding isAuthenticated, I do understand that people think it does something different, but the name is `isAuthenticated`, and not `isAuthorized`, so its name indicates it should be used to verify authentication, and not authorization. Having said that, happy to improve our documentation, what would you recommend?

As mentioned above, we considered dropping the concept of isAuthenticated altogether, and only offer a `user` and `tokens`. But this wouldnt neccesarily change much, as you can have a user, but not be authorized to call any API's.

**Comment by jimniels** [comment]
> in essence you are wrapping `getTokenSIlently` and could just forget about the concept of being authorized and just think in terms of "do I have a token to do what I want", and rely on our `getTokenSilently` method.

Yeah, this is a good point. Creating the idea of `isAuthorized` probably doesn't really need to happen. At the highest level of our app, where we protect routes, we can just call `getTokenSilently` and use that as our gatekeeping mechanism, because our use case is really simple: does this person have a token to talk to our API?

> happy to improve our documentation, what would you recommend?

For someone like me, who is just tasked with implementing auth and is by no means an auth expert, I (mistakenly) conflated the idea of being "authenticated" with being "authorized", which you more clearly described in this thread. I think some high-level mention of this in the docs could be useful as I would venture to guess it's easy for people not steeped in auth to conflate the two.

For example, [on this page under the heading "Evaluation the authentication state"](https://auth0.com/docs/quickstart/spa/vanillajs/01-login#evaluate-the-authentication-state), it says:

> you want to make sure anyone is able to visit the public page but not the page that is meant for authenticated users only, such as a settings panel or the user profile details. You can decide which content is available by hiding, disabling, or removing it if no user is currently logged in. You do so by checking the result of calling the `auth0Client.isAuthenticated()` method

When I first read this, I thought "ok cool, if there's a page with user-specific information that requires they be logged in to see, I can just check `isAuthenticated` to determine whether I should then call the API to show that". But where are you getting user-specific information? Very likely from your own API, which means you're calling `getTokenSilently`. I mistakenly assumed `isAuthenticated` = `true` guarantees `getTokenSilenty` will give back a token because, well, they're authenticated.

I suppose more succinctly, it was easy for me to think "user is authenticated" means "I can call my API to get their data".

For me personally, what would've helped was just some additional clarification in the docs that's really more educational than anything. Something that states what you stated in this thread: authentication !== authorization (maybe this kind of educational material already exists? if so, I didn't encounter it when implementing our version of auth0).

**Comment by frederikprijck** [comment]
I opened a PR to add something to the FAQ in the underlying SDK. It's called out in [the FAQ for auth0-react](https://github.com/auth0/auth0-react/blob/main/FAQ.md) to also read the FAQ for SPA-JS.

**Comment by Bartheleway** [comment]
> It's urgent that Auth0 clarify that because it cause a lot of confusion especially when your own documentation state that login should be offered only if `isAuthenticated` returns false :
> 
> * https://auth0.com/docs/quickstart/spa/vanillajs/01-login#restoring-login-state-with-social-providers
> * https://auth0.com/docs/quickstart/spa/vanillajs/interactive
> * https://github.com/auth0-samples/auth0-javascript-samples/blob/d45978c40db91b30ef81abcd9114f784aff5234b/01-Login/public/js/app.js#L62
> * https://auth0.github.io/auth0-spa-js/classes/Auth0Client.html#isAuthenticated

@frederikprijck could you give me an answer regarding that ? Every piece of code Auth0 is giving as example shows that `isAuthenticated` should only return true if the token is valid which is not quiet what you said before. Should we consider the documentation unclear / incorrect or the code is having an issue ?

**Comment by frederikprijck** [comment]
@Bartheleway can you point me to the docs that mention that about the token expiration vs considered being authenticated?

**Comment by Bartheleway** [comment]
Yes sure.

Here : https://auth0.com/docs/quickstart/spa/vanillajs/01-login#evaluate-the-authentication-state
You have a piece of code showing that login button should be enabled only if `isAuthenticated` returns true:
```javascript
const updateUI = async () => {
  const isAuthenticated = await auth0Client.isAuthenticated();

  document.getElementById("btn-logout").disabled = !isAuthenticated;
  document.getElementById("btn-login").disabled = isAuthenticated;
};
```
Let's say my token is expired but `isAuthenticated` returns true, what should I do ? I can't login again so my user would have no other choice than clicking on logout and then login.


Also here: https://auth0.com/docs/quickstart/spa/vanillajs/interactive#show-user-profile-information
You will find on the left side the code example:
```javascript
const profileElement = document.getElementById("profile");

if (isAuthenticated) {
  profileElement.style.display = "block";
  profileElement.innerHTML = `
          <p>${userProfile.name}</p>
          <img src="${userProfile.picture}" />
        `;
} else {
  profileElement.style.display = "none";
}
```
In this case, I will show personal information with an expired token, I don't think this is a good thing. I know it is just information publicly shared inside the token but still.

And here: https://auth0.github.io/auth0-spa-js/classes/Auth0Client.html#isAuthenticated
> Returns true if there's valid information stored, otherwise returns false.
This doesn't really explains what is `valid information`, it could be consistent session information or more a valid session (meaning token isn't expired).

**Comment by frederikprijck** [comment]
Thanks for sharing. All of that looks expected.

As mentioned in this thread, an access token being expired shouldnt impact the authentication state.

You can use getTokenSilently, it will allow you to get a new access token if the current is expired.

**Comment by Bartheleway** [comment]
So according to you, the method inform us at an IDP level that a valid session still exist but not at the application / API level.

In such situation, don't you think the SDK should provide a method to check if a valid session still exist for the application (could be a different method)? Maybe `checkSession` is a better candidate in such situation ?

**Comment by frederikprijck** [comment]
No, isAuthenticated shows if the user is authenticated in the application, not at the IdP or API level.

**Comment by Bartheleway** [comment]
Hi, so I tried to modify my implementation and using `checkSession` doesn't work in my case :(
Digging through the code of the SDK, I found that `checkSession` (or `getTokenSilently`) is checking the Auth0 session token expiration but not checking the application token expiration thus not refreshing my token when expired. I found a workaround calling `getTokenSilently` with a `cacheMode: 'off'` but it completly bypass the cache and call `/authorize` even if the token is still valid.

Basically I'm trying to do:
```javascript
async function isLogged () {
	let isLogged = await instance.isAuthenticated()

	if (isLogged) {
		try {
			await instance.getTokenSilently({
				cacheMode: 'off',
			})
		} catch () {
			isLogged = false
		}
	}

	return isLogged
}
```

**Comment by frederikprijck** [comment]
> Digging through the code of the SDK, I found that checkSession (or getTokenSilently) is checking the Auth0 session token expiration but not checking the application token expiration thus not refreshing my token when expired.

Can you show where that happens? As I am not sure I understand what you are trying to say.

> I found a workaround calling getTokenSilently with a cacheMode: 'off' but it completly bypass the cache and call /authorize even if the token is still valid.

That is expected to happen when you disable the cache, so the behavior seems correct.

Additionally, I am not sure I understand what you are trying to do, but let me try and elaborate what I think you want in a few situations:

- If you want to know if the user is **authenticated in the application**, use `isAuthenticated()`.
- If you want to know if the user is **authenticated at Auth0**, use `getTokenSilently({ cacheMode: 'off'})`. If you are no longer logged in at Auth0, it will throw a `login_required` error. **However, this only works when not using refresh tokens.** If you use refresh tokens, you are asking for `offline_access`, and are telling our SDK you want to disconnect from the Auth0 session. Also know this is not something I recommend to do often, as it bypasses our cache and you shouldnt need this.
- If you want to know if the user can call an API (and has a valid access token), use `getTokenSilently()` (note, no need to bypass the cache here)

Regarding the second bullet, and why I say you do not need it is the following:

- If you are not using refresh tokens, you should use `getTokenSilently()` (without disabling the cache). As long as the Access Token is not expired, it will use it from the cache. Once it's expired, it calls Auth0 (so it only calls it when necessary). If you are no longer logged in, it will throw a `login_required` and you know the user is no longer logged in to Auth0 (even better [our SDK will log the user out of the SDK in this case automatically](https://github.com/auth0/auth0-spa-js/blob/main/src/Auth0Client.ts#L927-L930)). The consequence here is that you will only know if you are still logged in once the access token is expired. However, as we recommend short-lived access tokens, this should be a matter of minutes and shouldnt cause any issues.
- If you are using refresh tokens, you are using `offline_access`, meaning you are talling our SDK you want to keep using Access Tokens even when no longer logged in to Auth0. That is by design and how refresh tokens work. In this case, there it makes no sense to ask Auth0 if the user is still logged in.

**Comment by cruzoscarjn** [comment]
Just to add something for anyone still struggling with this, define the auth provider without an specific audience and set the max_age to the same you want your "auth" to live.

The next time the app do any getSilentToken it will log out you and will set isAuthenticated as false
